src/haddock/gear/known_cns_errors.py

```diff
@@ -0,0 +1,177 @@
+        ),
+    "NCS-restraints error encountered: Improperly defined non-crystallographic symmetry": (  # noqa : E501
+        "Improperly defined non-crystallographic symmetry (NCS). "
+        "Check your symmetry restraints definition."
+        ),
+    "error in SYMMETRY potential, check NOE table": (
+        "Check your symmetry restraints definition."
+        ),
+    "exceeded allocation for NOE-restraints": (
+        "Too many distance restraints defined. "
+        "Try to reduce this number by checking your definition of active "
+        "and passive residues. "
+        "Make sure to filter those for solvent accessibility. "
+        "Or alternatively increase the nres parameter in the noe statements"
+        " in the relevant CNS scripts."
+        ),
+    "SELRPN error encountered: parsing error": (
+        "Check your restraint files."
+        ),
+    "PARSER error encountered: Encountered too many parsing errors": (
+        "Encountered too many parsing errors. "
+        "Check your input molecules and symmetry restraints."
+        ),
+    "XMREAD error encountered:  sectioning of map incompatible with resolution": (  # noqa : E501
+        "Check your EM map resolution and sectioning."
+        ),
+    "ALLHP error encountered: not enough memory available": (
+        "Too many distance restraints defined. "
+        "Try to reduce this number by checking your definition of active and "
+        "passive residues. "
+        "Make sure to filter those for solvent accessibility. "
+        "Try to decrease the size of your system where possible."
+        ),
+    "error encountered: missing SCATter definition for SELEcted atoms": (
+        "Unsupported atoms/molecules for cryo-EM restraints."
+        ),
+    "ROTMAT error encountered: rotation vector has zero length": (
+        "Check your input parameters and restraints. "
+        "Possibly try turning off the sampling of 180 degrees rotation."
+        )
+    }
+
+
+def find_cns_errors(cns_out_fpath: FilePath) -> Optional[KnownCNSError]:
+    """Detect if a known CNS error is in a cns.cnserr file.
+
+    Parameters
+    ----------
+    cns_out_fpath : FilePath -> Union[str, Path]
+        Path to the cns.cnserr file to check.
+
+    Returns
+    -------
+    Optional[KnownCNSError]
+        An exception for known CNS errors, with its hint on how to solve it!
+    """
+    # Check for file extension
+    if Path(cns_out_fpath).suffix == ".gz":
+        file_handle = gzip.open(cns_out_fpath, "rb")
+    else:
+        file_handle = open(cns_out_fpath, "rb")
+    try:
+        _find_cns_errors(file_handle, KNOWN_ERRORS, filepath=cns_out_fpath)
+    except KnownCNSError as err:
+        return err
+    else:
+        return None
+
+
+def _find_cns_errors(
+        file_handle: Union[gzip.GzipFile, BufferedReader],
+        known_errors: dict[str, str],
+        chunk_size: int = 4096,
+        filepath: FilePath = "",
+        ) -> None:
```

Comment:
Here you should also handle the situation in which there was no known error, maybe then raise a generic one directing people to one of the helper resources (manual/forum/etc)

Reply:
Good point, especially that now we known that files with the `.cnserr` will always contain an error.